util/pyclient/replica_specific_info.py

```diff
@@ -78,8 +88,28 @@ def num_matching_replies(self, matched_reply_key):
     def get_rsi_replies(self, matched_reply_key):
         if matched_reply_key not in self.replies.keys():
             return dict()
-        return self.replies[matched_reply_key]
+        rsi_replies = dict()
+        for k,v in self.replies[matched_reply_key].items():
+            rsi_replies[k] = v.get_rsi_data()
+        return rsi_replies
+
+    def get_all_replies(self):
+        replies = dict()
+        for seq_num, key in self.seq_nums.items():
+            replies[seq_num] = next(iter(self.replies[key].values()))
+        return replies
 
     def clear_replies(self):
         self.replies = dict()
 
+    def expects_seq_num(self, seq_num):
+        return seq_num in self.seq_nums.keys()
+
+    def has_quorum_on_all(self, required_replies):
+        has_quorum = True
+        for seq_num, key in self.seq_nums.items():
+            if key is None:
+                has_quorum = False
+            elif self.num_matching_replies(key) < required_replies:
+                has_quorum = False
```

Comment:
For a safer implementation, I'd actually reverse the logic here. Let's start by assuming `has_quorum = False` and then set it to `True` only in the relevant cases.

Reply:
We check all the expected messages if they have reached quorum and if any single one of them hasn't then we don't have quorum.